### PR TITLE
OPENEUROPA-3135: Add composite entity reference field functionality.

### PR DESCRIPTION
**Pull request by imanoleguskiza**
### Description

Add the possibility to flag entity reference fields as composite. Compositte fields will delete the referenced entity if no other entity is referencing it.
### Change log

- Added: Composite entity reference fields.
- Changed:
- Deprecated:
- Removed:
- Fixed:
- Security:

### Commands

```sh
[Insert commands here]

```

### CODE REVIEW

File: oe_content.module [hunk 1]
```diff
@@ -106,3 +109,128 @@ function oe_content_entity_base_field_info(EntityTypeInterface $entity_type) {
...
+  if (!in_array($field_config->getType(), $applicable_field_types)) {
+    return;
+  }
+  $form['composite'] = [
+    '#type' => 'radios',
+    '#title' => t('Composite field'),
+    '#description' => t('Making this field a composite field will delete any referenced entity if the referencing entity is deleted (only if the referenced entity is not referenced by some other entity).'),
+    '#default_value' => $field_config->getThirdPartySetting('oe_content', 'composite') ? 1 : 0,
+    '#options' => [
+      1 => t('Enabled'),
```
Comment (upchuk) [inline oe_content.module:131]: Why we not making this a boolean? As the ticket says: `A simple boolean will be enough`

File: oe_content.module [hunk 2]
```diff
@@ -106,3 +109,128 @@ function oe_content_entity_base_field_info(EntityTypeInterface $entity_type) {
...
+ *   The name of the entity type.
+ * @param \Drupal\field\FieldConfigInterface $field_config
+ *   The field config.
+ * @param array $form
+ *   The form array.
+ * @param \Drupal\Core\Form\FormStateInterface $form_state
+ *   The form state.
+ */
+function oe_content_form_field_config_form_builder(string $entity_type, FieldConfigInterface $field_config, array &$form, FormStateInterface $form_state): void {
+  if ($form_state->getValue('composite')) {
```
Comment (upchuk) [inline oe_content.module:155]: 6 lines that can be replaced with 

```
$field_config->setThirdPartySetting('oe_content', 'composite', (bool) $form_state->getValue('composite'));
```

File: oe_content.module [hunk 3]
```diff
@@ -106,3 +109,128 @@ function oe_content_entity_base_field_info(EntityTypeInterface $entity_type) {
...
+ * Implements hook_entity_delete().
+ *
+ * Delete any composite referenced entities when deleting an entity.
+ */
+function oe_content_entity_delete(EntityInterface $entity): void {
+  if (!$entity instanceof \Drupal\Core\Entity\FieldableEntityInterface) {
+    return;
+  }
+  /** @var \Drupal\Core\Field\FieldDefinitionInterface[] $field_definitions */
+  $field_definitions = \Drupal::service('entity_field.manager')->getFieldDefinitions($entity->getEntityTypeId(), $entity->bundle());
```
Comment (upchuk) [inline oe_content.module:173]: Let's move all this logic that looks for fields and does the delete and checks if it's not referennced already do a standalone service. Let's have something like: 

`CompositeReferenceFieldManager::onDelete()` getting the field value. So you check in the hook if it's the correct field type, then you defer to the service. And in the service you get the references, check them and delete them.

For the method that checks if we can delete, let's name it and make it generic: `getReferencingEntities()` and it should return info about the referencing entities. We might use this elsewhere as well, or someone would.

File: oe_content.module [hunk 4]
```diff
@@ -106,3 +109,128 @@ function oe_content_entity_base_field_info(EntityTypeInterface $entity_type) {
...
+ */
+function oe_content_referenced_entity_deletable(EntityInterface $entity): bool {
+  $entity_type_manager = \Drupal::entityTypeManager();
+
+  // Getting fields which could have references to the entity
+  // we are trying to delete.
+  $fields = $entity_type_manager->getStorage('field_config')->loadByProperties([
+    'field_type' => 'entity_reference',
+  ]);
+  $fields = array_merge($fields, $entity_type_manager->getStorage('field_config')->loadByProperties([
```
Comment (upchuk) [inline oe_content.module:208]: lol, why don't you just loadByProperties(['field_type', ['entity_reference', 'entity_reference_revisions']])?

File: tests/Kernel/CompositeFieldsTest.php [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    'user',
+  ];
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function setUp(): void {
+    parent::setUp();
+
+    $this->installSchema('user', 'users_data');
```
Comment (upchuk) [inline tests/Kernel/CompositeFieldsTest.php:36]: Do we need users_data?

**Review by upchuk — CHANGES_REQUESTED** [review]
We are missing form alter coverage. So we need a functional test as well.

File: oe_content.module (hunk 4 above)
Comment (imanoleguskiza) [reply oe_content.module:208]: I tried that but loading field types by properties doesn't use the IN sql selector, so it ends up returning nothing....

File: oe_content.module [hunk 6]
```diff
@@ -106,3 +109,128 @@ function oe_content_entity_base_field_info(EntityTypeInterface $entity_type) {
...
+  /** @var \Drupal\field\FieldConfigInterface $field_config */
+  $field_config = $form_state->getFormObject()->getEntity();
+  $applicable_field_types = ['entity_reference', 'entity_reference_revisions'];
+  if (!in_array($field_config->getType(), $applicable_field_types)) {
+    return;
+  }
+  $form['composite'] = [
+    '#type' => 'radios',
+    '#title' => t('Composite field'),
+    '#description' => t('Making this field a composite field will delete any referenced entity if the referencing entity is deleted (only if the referenced entity is not referenced by some other entity).'),
```
[the remark this answers is not in the document]
Comment (imanoleguskiza) [reply oe_content.module]: Fixed.

File: oe_content.module (hunk 1 above)
Comment (imanoleguskiza) [reply oe_content.module:131]: If make the options boolean then for some reason they don't get checked properly once they are rendered.

File: oe_content.module (hunk 2 above)
Comment (imanoleguskiza) [reply oe_content.module]: Fixed, much better indeed.

File: oe_content.module (hunk 3 above)
Comment (imanoleguskiza) [reply oe_content.module:168]: Done, please check if the service is what you where envisioning.

File: tests/Kernel/CompositeFieldsTest.php (hunk 5 above)
Comment (imanoleguskiza) [reply tests/Kernel/CompositeFieldsTest.php]: Removed.

File: oe_content.module [hunk 7]
```diff
@@ -106,3 +109,70 @@ function oe_content_entity_base_field_info(EntityTypeInterface $entity_type) {
...
+  /** @var \Drupal\field\FieldConfigInterface $field_config */
+  $field_config = $form_state->getFormObject()->getEntity();
+  $applicable_field_types = ['entity_reference', 'entity_reference_revisions'];
+  if (!in_array($field_config->getType(), $applicable_field_types)) {
+    return;
+  }
+  $form['composite'] = [
+    '#type' => 'radios',
+    '#title' => t('Composite field'),
+    '#description' => t('Composite reference fields ensure that referenced entities get deleted when the referencing entity is deleted..'),
```
Comment (upchuk) [inline oe_content.module:128]: Extra period at the end...

File: oe_content.module (hunk 1 above)
Comment (upchuk) [reply oe_content.module:131]: What?

File: src/CompositeReferenceFieldManager.php [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  public function getReferencingEntities(EntityInterface $entity): array {
+    $referencing_entities = [];
+    // Getting all fields which could have references to the given entity.
+    $fields = $this->entityTypeManager->getStorage('field_config')->loadByProperties([
+      'field_type' => 'entity_reference',
+    ]);
+    $fields = array_merge($fields, $this->entityTypeManager->getStorage('field_config')->loadByProperties([
```
Comment (upchuk) [inline src/CompositeReferenceFieldManager.php:45]: Try loading via query->condition with IN to see if you can merge them into one query.

**Review by upchuk — CHANGES_REQUESTED** [review]

File: src/CompositeReferenceFieldManager.php [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+  /**
+   * {@inheritdoc}
+   */
+  public function deleteCompositeReferences(EntityInterface $entity, FieldDefinitionInterface $field_definition): void {
+    if ($field_definition instanceof FieldConfigInterface && $field_definition->getThirdPartySetting('oe_content', 'composite', FALSE)) {
+      $referenced_entities = $entity->get($field_definition->getName())->referencedEntities();
+      /** @var \Drupal\Core\Entity\EntityInterface $referenced_entity */
+      foreach ($referenced_entities as $referenced_entity) {
+        if ($referenced_entity->id() !== $entity->id() && empty($this->getReferencingEntities($referenced_entity))) {
```
[the remark this answers is not in the document]
Comment (imanoleguskiza) [reply src/CompositeReferenceFieldManager.php:81]: AFAIK there is no restriction in place so an  entity could reference itself.

File: src/CompositeReferenceFieldManager.php (hunk 8 above)
Comment (imanoleguskiza) [reply src/CompositeReferenceFieldManager.php]: It worked!

File: tests/Kernel/CompositeFieldsTest.php [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+          'target_bundles' => [
+            'test_ct' => 'test_ct',
+          ],
+        ],
+      ],
+    ]);
+    $entity_reference_field->setThirdPartySetting('oe_content', 'composite', FALSE);
+    $entity_reference_field->save();
+
+    $node_storage = $entity_type_manager->getStorage('node');
```
[the remark this answers is not in the document]
Comment (imanoleguskiza) [reply tests/Kernel/CompositeFieldsTest.php:69]: I have added more comprehensive comments and added coverage for entity reference revisions.

File: tests/Traits/CompositeReferenceTestTrait.php [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+      ]);
+      if ($revision) {
+        $reference_field_storage->set('type', 'entity_reference_revisions');
+      }
+      $reference_field_storage->save();
+    }
+    if ($reference_field = FieldConfig::loadByName($entity_type, $bundle, $field_name)) {
+      return $reference_field;
+    }
+    $reference_field = FieldConfig::create([
```
Comment (upchuk) [inline tests/Traits/CompositeReferenceTestTrait.php:64]: Do a check if it doesn't exist already as well.

File: tests/Traits/CompositeReferenceTestTrait.php [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    return $reference_field;
+  }
+
+  /**
+   * Entity reference field definition provider.
+   *
+   * @return array
+   *   An array of entity reference definitions.
+   */
+  public function compositeFieldDefinitions(): array {
```
Comment (upchuk) [inline tests/Traits/CompositeReferenceTestTrait.php:84]: Does not belong in the trait, nor should we use data providers

File: tests/Traits/CompositeReferenceTestTrait.php [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+      'entity_type' => $entity_type,
+      'bundle' => $bundle,
+      'label' => $field_label,
+      'settings' => [
+        'handler' => $selection_handler,
+        'handler_settings' => $selection_handler_settings,
+      ],
+    ]);
+    $reference_field->save();
+    return $reference_field;
```
Comment (upchuk) [inline tests/Traits/CompositeReferenceTestTrait.php:75]: Let's have please (for me) an empty line before the return. Jumps in my eye and it hurts me :)

File: tests/Kernel/CompositeFieldsTest.php [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+
+use Drupal\Tests\oe_content\Traits\CompositeReferenceTestTrait;
+use Drupal\Tests\rdf_entity\Kernel\RdfKernelTestBase;
+
+/**
+ * Tests composite reference fields.
+ */
+class CompositeFieldsTest extends RdfKernelTestBase {
+
+  use CompositeReferenceTestTrait;
```
Comment (upchuk) [inline tests/Kernel/CompositeFieldsTest.php:15]: Empty line afteer the use statement

File: tests/Kernel/CompositeFieldsTest.php [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+  /**
+   * Modules to enable.
```
Comment (upchuk) [inline tests/Kernel/CompositeFieldsTest.php:17]: {@inheritdoc}

File: tests/Functional/CompositeReferenceConfigurationTest.php [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+use Drupal\Tests\oe_content\Traits\CompositeReferenceTestTrait;
+
+/**
+ * Functional tests for composite reference configuration forms.
+ *
+ * @package Drupal\Tests\oe_content\Functional
+ */
+class CompositeReferenceConfigurationTest extends BrowserTestBase {
+
+  use CompositeReferenceTestTrait;
```
Comment (upchuk) [inline tests/Functional/CompositeReferenceConfigurationTest.php:18]: Line after use statement

**Review by upchuk — CHANGES_REQUESTED** [review]

File: tests/Traits/CompositeReferenceTestTrait.php (hunk 11 above)
Comment (imanoleguskiza) [reply tests/Traits/CompositeReferenceTestTrait.php:64]: I do right above :)

**Review by upchuk — DISMISSED** [review]

**Comment by imanoleguskiza** [comment]
Outdatted PR

**Pull request closed without merging**